Rpackage: Extension methods should know their package

jecisc opened the pull request:
Now extension methods know directly their package which will speed up a lot of things. (See bellow for some examples) Those speed up will allow me to remove multiple caches in RPackage and simplify a lot the code of RPackage! 

Another change this will enable is for the methods to be able to return their package even after they are removed from the system. The next step to enable this is for the class to still know their package after removal (for non extension methods). This is planed for the next few weeks! This will also simplify and speedup more things in the future!

I'm kinda excited by all the speedup and simplifications this will allow me to do and the code to do that is not that complex.

Benches:

```st
allMethods := CompiledMethod allInstances.

[ allMethods collect: #isExtension] timeToRun. 

Before: "0:00:00:00.358"

After: "0:00:00:00.005" 

[ allMethods collect: #package ] timeToRun. 

Before: "0:00:00:00.455"

After: "0:00:00:00.195"  
```

I might add more things to this PR to start removing the caches

Edit: I added the removal of a cache. Here are the benches:

```st
allClasses := Smalltalk allClassesAndTraits.

[ allClasses collect: #extendingPackages ] timeToRun.

Before: "0:00:00:00.009"

After: "0:00:00:00.038"
```

This is slower but I think it is worth it because of the reduction of the complexity of the code and also this code is not called a lot compared to #package or #isExtension

jecisc committed:
Add a property to the methods to know their extension package

jecisc committed:
Fix test

jecisc committed:
Remove classExtendingPackagesMapping cache

jecisc commented:
I've check the generated image and some methods seems to be considered as defined methods while they are extension methods :( I need to find why

jecisc commented:
It seems to come from recompilation, let's fix that

jecisc committed:
Move the persistance of #extensionMethod property

astares commented:
Cool

jecisc commented:
I can confirm my fix produces an image that seems right

Ducasse commented:
Cyril now where is stored the package of a method? In the method property I imagine.

Ducasse commented:
A test I would like to see in particular for the tags is to open calypso on Bloc core package

jecisc commented:
I save it in the method properties indeed.
I checked and we do not have that much extension methods so it will not add too much memory. And also, with the caches I’ll be able to remove, we will probably gain memory in the end :)

To test the opening on Calypso on Bloc we would need Bloc in the image first

jecisc commented:
I think I found another bug with this. If I remove an ivar from a class, the property is lost. I think this is because when migrating the methods from the old to the new class there is no prior method present to migrate the property. I'll fix that

jecisc committed:
Fix bug during recompilation of classes that removed the #extensionPackage property of methods

jecisc committed:
Add regression test

jecisc committed:
Add another regression test

jecisc commented:
I just added a few tests for the problems I found and fix

jecisc commented:
It seems that we have a test crashing but nothing to help in the logs :(

jecisc committed:
Improve logs while running tests

The goal of the change is to improve logs while running tests.

It prints when it starts, when it ends and how long it took.

Also it now uses the Transcript instead of Stdout to be sure the logs are in the right order (because transcript is buffered). Also I flush often because if a test crashes the image while the buffer of the transcript is not flushed. we cannot know when the problem occured.

jecisc committed:
Merge branch 'misc/improve-running-tests-logs' into packages/improve-methods-management

jecisc committed:
Fix return of the report

jecisc committed:
Merge branch 'misc/improve-running-tests-logs' into packages/improve-methods-management

Ducasse commented:
Once this is integrated I will load Bloc in P12. Bloc works well in P11 so it should work on P12.

jecisc committed:
Fix bug with OCCallMethodProxy

Those classes acting as methods with the #run:with:in: are just plain HORRIBLE! Each time we change something in the system related to methods those classes are breaking.  :(

jecisc commented:
So what was breaking now it the classes that acts like methods with the method #run:with:in: 

I find this system horrible because we have to duplicate part of the compiled method API in a lot of places and each time we change something in the methods management something breaks :(

jecisc commented:
This is now working! :)

Ducasse commented:
I agree with you and we should remove the things that are bizarre and unefficient. 
If you have a list of the API that is duplicated please log it so that we can clean it.
